feat(rpm) Added RPM versioning functionality

laundry-96 opened the pull request:
## Changes

Added RPM functionality with the help of @JamieMagee.
I had to reset the changes due to conflicts, but Jamie did most of the work so mentioning them for credit.

## Context

Closes  #20465

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [X] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [X] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

viceice reviewed (commented):
missing coverage

laundry-96 commented:
@viceice How can I fix the coverage check?
I have read the development docs, but I only see that I need 100% coverage, and checking coverage/index.html that I need to cover some more branches (I'm assuming if statements).

If you have any more docs or some other code I can reference, that would be very helpful! And thank you so much for taking time out of your weekend to check :)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/renovatebot/renovate?pullRequest=26340) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/renovatebot/renovate?pullRequest=26340) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: laundry-96<br/>:x: Austin DeLauney<br/><hr/>**Austin DeLauney** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/renovatebot/renovate?pullRequest=26340) it.</sub>

laundry-96 commented:
There were some tests in index.spec.ts that I fixed by finding the correct value by using rpm.labelCompare in Python.

I also fixed the issues coverage, and added more functionality that is supported by RPM for full coverage.
I have learned more than I ever wanted about RPM versioning, and I now realize why no one has implemented this yet 🥲

viceice commented on lib/modules/versioning/rpm/index.ts:
```diff
@@ -0,0 +1,290 @@
+        matchv2 = matchv2?.replace(leadingZerosPattern, '');
+
+        //We clearly have a number here, so return which is greater
+        const num1 = Number(matchv1);
+        const num2 = Number(matchv2);
```
```suggestion
        const num1 = Number.parseInt(matchv1, 10);
        const num2 = Number.parseInt(matchv2, 10);
```
we should be explicit here

viceice replied:
use `0` if string is empty. a better solution would be to leave one zero if the string only contains zeros.

laundry-96 replied:
Made a better regex :)

viceice reviewed (commented):
otherwise LGTM

laundry-96 commented:
@viceice Accepted! Thank you :)

viceice commented on lib/modules/versioning/rpm/index.ts:
```diff
@@ -0,0 +1,290 @@
+      return 0;
+    }
+
+    const matchesv1 =
+      (v1.match(alphaNumPattern) as Array<string>) ?? ([] as Array<string>);
+    const matchesv2 =
+      (v2.match(alphaNumPattern) as Array<string>) ?? ([] as Array<string>);
```
```suggestion
    const matchesv1 = v1.match(alphaNumPattern) ?? [];
    const matchesv2 = v2.match(alphaNumPattern) ?? [];
```
ups, I don't think we need those casts

viceice commented on lib/modules/versioning/rpm/index.ts:
```diff
@@ -0,0 +1,290 @@
+    const matches = Math.min(matchesv1.length, matchesv2.length);
+
+    for (let i = 0; i < matches; i++) {
+      let matchv1 = matchesv1?.[i];
+      let matchv2 = matchesv2?.[i];
```
```suggestion
      let matchv1 = matchesv1[i];
      let matchv2 = matchesv2[i];
```
those are never null or undefined because of the fallback to empty array above

viceice commented on lib/modules/versioning/rpm/index.ts:
```diff
@@ -0,0 +1,290 @@
+    }
+
+    // If there is a tilde in a segment past the minimum number of segments, find it
+    if (matchesv1.length > matches && matchesv1?.[matches][0] === '~') {
```
```suggestion
    if (matchesv1.length > matches && matchesv1[matches][0] === '~') {
```

viceice commented on lib/modules/versioning/rpm/index.ts:
```diff
@@ -0,0 +1,290 @@
+      return -1;
+    }
+
+    if (matchesv2.length > matches && matchesv2?.[matches][0] === '~') {
```
```suggestion
    if (matchesv2.length > matches && matchesv2[matches][0] === '~') {
```

viceice reviewed (commented):
sorry, found some more.

laundry-96 commented:
@viceice The Number.parseInt doesn't work because the lines above remove leading zeros. If a string is '0' it turns into '', which that function returns NaN.

viceice reviewed (commented):
see comment

viceice commented on lib/modules/versioning/rpm/index.ts:
```diff
@@ -0,0 +1,288 @@
+          return 1;
+        }
+
+        // trim leading zeros
+        matchv1 = matchv1?.replace(leadingZerosPattern, '');
+        matchv2 = matchv2?.replace(leadingZerosPattern, '');
```
do we really need to trim leading zeros for parseInt ?

viceice commented on lib/modules/versioning/rpm/index.ts:
```diff
@@ -0,0 +1,283 @@
+        }
+
+        //We clearly have a number here, so return which is greater
+        const num1 = Number.parseInt(matchv1, 10);
+        const num2 = Number.parseInt(matchv2, 10);
+
+        const result = num1 - num2;
```
```suggestion
        const result = matchv1.localCompare(matchv2, undefined, { numeric: true });
```
isn't this the same result but much faster?

laundry-96 replied:
I'm not the best with javascript, so you're probably right haha.

viceice commented:
> How can this manager be used? Didn't find documentation for it. Thanks.

please open a new discussion.

this is a versioning, not a manager 😉
